Public/Src/Tools/DropDaemon/Program.cs

```diff
@@ -518,77 +519,103 @@ private static void SetupThreadPoolAndServicePoint()
+            if (directoryPaths.Length != directoryIds.Length || directoryPaths.Length != directoryDropPaths.Length || directoryPaths.Length != directoryFilters.Length)
+            {
+                return new IpcResult(
+                    IpcResultStatus.GenericError,
+                    Inv(
+                        "Directory counts don't match: #directories = {0}, #directoryIds = {1}, #dropPaths = {2}, #directoryFilters = {3}",
+                        directoryPaths.Length, directoryIds.Length, directoryDropPaths.Length, directoryFilters.Length));
+            }
 
-                    result = await AddDropItemsAsync(daemon, finalListOfFilesToUpload);
-                }
+            (Regex[] initializedFilters, string filterInitError) = InitializeDirectoryFilters(directoryFilters);
+            if (filterInitError != null)
+            {
+                return new IpcResult(IpcResultStatus.ExecutionError, filterInitError);
+            }
 
-                daemon.Logger.Verbose("[ADDARTIFACTS] " + result);
-                return result;
-            });
+            var dropFileItemsKeyedByIsAbsent = Enumerable
+                .Range(0, files.Length)
+                .Select(i => new DropItemForBuildXLFile(
+                    daemon.ApiClient,
+                    chunkDedup: conf.Get(EnableChunkDedup),
+                    filePath: files[i],
+                    fileId: fileIds[i],
+                    fileContentInfo: FileContentInfo.Parse(hashes[i]),
+                    relativeDropPath: dropPaths[i])).ToLookup(f => WellKnownContentHashUtilities.IsAbsentFileHash(f.Hash));
+
+            // If a user specified a particular file to be added to drop, this file must be a part of drop.
+            // The missing files will not get into the drop, so we emit an error.
+            if (dropFileItemsKeyedByIsAbsent[true].Any())
+            {
+                return new IpcResult(
+                    IpcResultStatus.InvalidInput, 
+                    Inv("The following files are missing, but they are a part of the drop command:{0}{1}", 
+                        Environment.NewLine,
+                        string.Join(Environment.NewLine, dropFileItemsKeyedByIsAbsent[true])));
+            }
+
+            (IEnumerable<DropItemForBuildXLFile> dropDirectoryMemberItems, string error) = await CreateDropItemsForDirectoriesAsync(
+                conf,
+                daemon,
+                directoryPaths,
+                directoryIds,
+                directoryDropPaths,
+                initializedFilters);
+
+            if (error != null)
+            {
+                return new IpcResult(IpcResultStatus.ExecutionError, error);
+            }
+
+            var groupedDirectoriesContent = dropDirectoryMemberItems.ToLookup(f => WellKnownContentHashUtilities.IsAbsentFileHash(f.Hash));
+
+            // we allow missing files inside of directories only if those files are output files (e.g., optional or temporary files) 

```

Comment:
>we allow missing files inside of directories only if those files are output files [](start = 15, length = 81)

As far as I understand from yesterday's discussion,  we only want to do this only for opaque (or shared opaque) directories. For partially or fully sealed directories where users explicitly specify the members of those directories, inclusion of absent files should trigger errors.

Reply:
I'm going to approve this PR. But I still think that the work is incomplete. For partially sealed and fully sealed directories, where users can specify the files explicitly, drop should fail if those files are absent.

It's up to you to complete it or not. Also, I think adding a server API to know the seal directory kind is not difficult.